src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs

```diff
@@ -1596,7 +1596,9 @@ await PreviewBuildIntegratedProjectActionsAsync(buildIntegratedProject, actions,
             INuGetProjectContext nuGetProjectContext,
             IReadOnlyCollection<SourceRepository> activeSources,
             VersionRange versionRange,
-            CancellationToken token)
+            CancellationToken token,
+            string newMappingID = null,
+            string newMappingSource = null)
```

Comment:
Firstly, calling this API with the extra parameters doesn't appear to be in this PR, so I assume this PR is setting up the public API for other work which will call it.

However, I have two "bigger picture" thoughts:

1. If the customer is installing a package from a private feed, which has dependencies that come from a public feed, is this a scenario that's supposed to work?
2. Do we think in the future it will be possible to install more than 1 package in a single action?
    * Currently it's possible in PM UI's upgrade tab to update more than one package. But unless the new version changes dependencies, an update generally shouldn't need a different package source mapping.
    * But I'm wondering if a future improvement to PM UI might be to allow installing multiple packages "at once", to reduce waiting time (install package 1, wait for it to finish, install package 2, wait for it to finish).

I'm wondering if "needing more than 1 package source mapping" is an "obvious" future need, and we can avoid future public API changes if we do it now, even if we only call it with 1 package id/source per call initially.

Reply:
Discussed a bit offline. The Preview restore accepts 1 package ID. I'm not sure I see how it supports batching multiple actions or if this PR should be the thing that implements such batching.
Seeking more clarification...

Reply:
Github not showing me all the method parameters confused me. I thought this was one of the methods that took `List<NuGetAction>`. I know the method that has all the logic for install & uninstall is like that, so I made a bad assumption, sorry.

The first point, about a dependency that lives on a different package source stands. In fact, I'm now wondering if this feature will work for packages that have dependencies. If I try to install PackageA, which depende on PackageB, then this method will be called with PackageA, Source1 as parameters, but as far as I can tell, nothing will set up the source mapping for PackageB. Of course, NuGet won't know about PackageB until it resolves the graph, which happens as part of the preview restore.  Anyway, all I've managed to do is convince myself this is a difficult feature to automate in a GUI.

Reply:
Yes, all of the dependencies will be source mapped. The UIActionEngine is taking the addedPackages, which is everything Restore deemed as necessary to add as part of this install. I then diff that with existing mappings and only write mappings that aren't currently satisfied by another mapping.

Added packages are analyzed here: https://github.com/NuGet/NuGet.Client/pull/5109/files/e93884166bf77bc495041788db16dc9efd9b687e#diff-ac2848ccdf455d1a83de39b70803b630e3b17f94534b18d15aa8fcf6c8191921R484